utils/tlv/tlv.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/sys/util.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <tlv.h>
+
+static tlv_header bytes_to_header(uint8_t data[4])
+{
+	return (tlv_header){ .type = ((data[0] << 8) + data[1]),
+			     .size = { .padding = data[2], .size = data[3] } };
+}
+
+static void header_to_bytes(tlv_header header, uint8_t data[4])
+{
+	data[0] = header.type >> 8;
+	data[1] = header.type & 0xFF;
+	data[2] = header.size.padding;
+	data[3] = header.size.size;
+}
+
+bool tlv_lookup(tlv_ctx *ctx, tlv_type type, tlv_header *lookup_data)
+{
+	for (uint32_t offset = ctx->first_entry_offset; offset <= ctx->last_valid_offset;) {
+		uint8_t header_raw[4];
+		ctx->storage.read(ctx->storage.ctx, offset, header_raw, sizeof(header_raw));
+		tlv_header header = bytes_to_header(header_raw);
+		offset += sizeof(header);
+		if (header.type != type) {
+			offset += header.size.size + header.size.padding;
+			continue;
+		} else {
+			*lookup_data = header;
+			return true;
+		}
+	}
+	return false; // NOT FOUND
+}
+
+bool tlv_read(tlv_ctx *ctx, tlv_type type, uint8_t *data, uint16_t data_size)
+{
+	for (uint32_t offset = ctx->first_entry_offset; offset <= ctx->last_valid_offset;) {
+		uint8_t header_raw[4];
+		ctx->storage.read(ctx->storage.ctx, offset, header_raw, sizeof(header_raw));
+		tlv_header header = bytes_to_header(header_raw);
+		offset += sizeof(header_raw);
+		if (header.type != type) {
+			offset += header.size.size + header.size.padding;
+			continue;
+		} else {
+			if (data_size <= header.size.size) {
+				ctx->storage.read(ctx->storage.ctx, offset, (uint8_t *)&data,
+						  data_size);
+				return true;
+			} else {
+				return false; // requested to read more than it is in storage.
+			}
+		}
+	}
+	return false; // NOT FOUND
+}
+
+bool tlv_write(tlv_ctx *ctx, tlv_type type, uint8_t *data, uint16_t data_size)
+{
+	if (ctx->last_valid_offset < (ctx->next_free_offset + sizeof(tlv_header) + data_size +
+				      CALCULATE_PADDING(data_size))) {
```

Comment:
the macro might be replaced with the [aligned attribute](https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Type-Attributes.html) ?

Reply:
I can not used alligned attribute, because I allow arbitrary size of the payload, and alligned attribute affects only address location in RAM, not offset in FLASH